core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -561,6 +593,234 @@ public UpdateNode plan(Update node)
+
+        RelationPlan target = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, recursiveSubqueries)
+                .process(merge.getTarget());
+
+        Assignments.Builder projections = Assignments.builder();
+        projections.putIdentities(target.getRoot().getOutputSymbols());
+
+        Symbol presentColumn = symbolAllocator.newSymbol("present", BOOLEAN);
+        projections.put(presentColumn, TRUE_LITERAL);
+
+        target = new RelationPlan(
+                new ProjectNode(idAllocator.getNextId(), target.getRoot(), projections.build()),
+                mergeAnalysis.getTargetTableScope(),
+                target.getFieldMappings(),
+                outerContext);
+
+        RelationPlan source = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, recursiveSubqueries)
+                .process(merge.getSource());
+
+        RelationPlan joinPlan = new RelationPlanner(analysis, symbolAllocator, idAllocator, lambdaDeclarationToSymbolMap, metadata, outerContext, session, recursiveSubqueries)
+                .planJoin(merge.getPredicate(), Join.Type.RIGHT, mergeAnalysis.getJoinScope(), target, source, analysis.getSubqueries(merge));
+
+        PlanBuilder subPlan = newPlanBuilder(joinPlan, analysis, lambdaDeclarationToSymbolMap);
+
+        ImmutableList.Builder<WhenClause> whenClauses = ImmutableList.builder();
+        for (int caseNumber = 0; caseNumber < merge.getMergeCases().size(); caseNumber++) {
+            MergeCase mergeCase = merge.getMergeCases().get(caseNumber);
+
+            if (mergeCase.getExpression().isPresent()) {
+                subPlan = subqueryPlanner.handleSubqueries(subPlan, mergeCase.getExpression().get(), analysis.getSubqueries(mergeCase));
+            }
+
+            ImmutableList.Builder<Expression> rowBuilder = ImmutableList.builder();
+            List<ColumnHandle> mergeCaseSetColumns = mergeCaseColumnsHandles.get(caseNumber);
+            for (ColumnHandle dataColumnHandle : mergeAnalysis.getDataColumnHandles()) {
+                int index = mergeCaseSetColumns.indexOf(dataColumnHandle);
+                if (index >= 0) {
+                    Expression setExpression = mergeCase.getSetExpressions().get(index);
+                    Type type = analysis.getCoercion(setExpression);
+                    if (type != null) {
+                        setExpression = new Cast(setExpression, toSqlType(type));
+                    }
+                    rowBuilder.add(subPlan.rewrite(setExpression));
+                }
+                else {
+                    Integer fieldNumber = requireNonNull(mergeAnalysis.getColumnHandleFieldNumbers().get(dataColumnHandle), "Field number for ColumnHandle is null");
+                    rowBuilder.add(target.getFieldMappings().get(fieldNumber).toSymbolReference());
```

Comment:
It depends on the `RowChangeParadigm`.  As mentioned above, all columns of the table will be written for the DeleteAndInsertRowChangeProcessor used by Hive.  For JDBC will want a new RowChangeProcessor that not only writes changed columns but also that sorts the rows by merge case number, so that it can perform batched updates.

Reply:
It depends on the `RowChangeParadigm`.  As mentioned above, all columns of the table will be written for the DeleteAndInsertRowChangeProcessor used by Hive.  For JDBC will want a new RowChangeProcessor that not only writes changed columns but also that sorts the rows by merge case number, so that it can perform batched updates.